lib/ocamlorg_web/graphql.ml

```diff
@@ -6,59 +6,125 @@ type package_info =
   ; constraints : string option
   }
 
-type packages_result =
+type packages_success =
   { total_packages : int
   ; packages : Package.t list
   }
 
-let starts_with s1 s2 =
-  let len1 = String.length s1 in
-  if len1 > String.length s2 then
-    false
-  else
-    let s1 = String.lowercase_ascii s1 in
-    let s2 = String.lowercase_ascii s2 in
-    String.(equal (sub s2 0 (length s1)) s1)
+type packages_result = (packages_success, [ `Msg of string ]) result
 
-let is_package s1 s2 =
-  let len1 = String.length s1 in
-  if len1 <> String.length s2 then
-    false
-  else
-    let s1 = String.lowercase_ascii s1 in
-    let s2 = String.lowercase_ascii s2 in
-    String.(equal s2 s1)
+type package_success = { package : Package.t }
 
-let get_packages_result total_packages offset limit startswith packages =
-  match startswith with
-  | None ->
-    let packages =
-      List.filteri (fun i _ -> offset <= i && i < offset + limit) packages
-    in
-    { total_packages; packages }
-  | Some filter ->
-    let packages =
-      List.filteri
-        (fun i package ->
-          offset <= i
-          && i < offset + limit
-          && starts_with filter (Package.Name.to_string (Package.name package)))
-        packages
-    in
-    { total_packages; packages }
-
-let get_single_package all_packages name =
-  List.find_opt
-    (fun package ->
-      is_package name (Package.Name.to_string (Package.name package)))
-    all_packages
+type package_result = (package_success, [ `Msg of string ]) result
 
 let get_info info =
   List.map
     (fun (name, constraints) ->
       { name = Package.Name.to_string name; constraints })
     info
 
+let is_in_range current_version from_version upto_version =
+  current_version >= from_version && current_version <= upto_version
+
+let is_valid_params limit offset total_packages =
+  let is_valid_limit = limit >= 0 && limit <= total_packages in
+  let is_valid_offset = offset >= 0 && offset <= total_packages in
+  let res = is_valid_limit in
+  match res with
+  | false ->
+    "wrong_limit"
+  | _ ->
+    let res = is_valid_offset in
+    (match res with false -> "wrong_offset" | _ -> "true")
```

Comment:
> Have you considered returning a custom variant type instead of a string? What do you think about that?

It's a bit unclear what you meant here but if you can drop an example, I will appreciate it

Reply:
> Why do you have limit <= total_packages? I think the highest possible limit is total_packages - offset (+/- 1), isn't it? It might also be an option to assume that it's ok to exceed the highest possible limit.

Yea, I think it is better as `limit  > total_packages` as you specified in your if statement. Thank you

Reply:
> It's a bit unclear what you meant here but if you can drop an example, I will appreciate it

I mean something along the lines
```
type params_validity = Valid_params | Wrong_limit | Wrong_offset
let is_valid_params limit offset total_packages =
  if limit < 0 || limit > total_packages then
    Wrong_limit
  else if offset < 0 || offset > total_packages then
    Wrong_offset
  else
    Valid_params
```

Reply:
> Yea, I think it is better as limit > total_packages as you specified in your if statement. Thank you

What I meant was that, if I understand the code below correctly, the highest possible limit is `total_packages - offset` (I might be off by 1 here), not `total_packages`. So I was suggesting either to change that bound to
```
limit > total_packages - offset (+/- 1)
```
or to completely leave it out since it might also make sense to allow a higher limit in the query and in that case to return all packages from the offset on.

Reply:
Offset will not always be 0. But I subtracted 1 from total_packages directly since it starts counting from 0. Thank you for pointing that out

Reply:
What I mean is that the upper bound for `limit` should either be 
```
limit > total_packages - offset (+/- 1)
```
or left out completely. I don't think it should be `limit > total_packages` or `limit > total_packages - 1` or anything like that. Do you understand why I'm saying that?

Reply:
Also, what do you think about the [suggestion about the custom return type](https://github.com/ocaml/v3.ocaml.org-server/pull/78#discussion_r694727416)?

Reply:
Hi @pitag-ha please do you mean something like this?
```
  let is_valid_params limit offset total_packages =
  if limit > total_packages - offset (+/- 1) then
    "wrong_limit"
  else if offset < 0 || offset > total_packages - 1 then
    "wrong_offset"
  else
    "true"
  ```

Reply:
> Also, what do you think about the suggestion about the custom return type?

let me try it out now

Reply:
I meant
```
if limit < 1 || limit > total_packages - offset (+/- 1) then
    Wrong_limit
```
or simply
```
if limit < 1 then
    Wrong_limit
```

Reply:
Thanks for the changes!